File: benchmarks/tsb/bench_ewm_adjust.ts
/**
 * Benchmark: EWM with adjust=false — IIR-based exponential weighted mean vs default adjust=true on 100k Series.
 * Outputs JSON: {"function": "ewm_adjust", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

const data = Array.from({ length: SIZE }, (_, i) => Math.sin(i * 0.01) * 100);
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  s.ewm({ alpha: 0.3, adjust: false }).mean();
  s.ewm({ alpha: 0.3, adjust: true }).mean();
  s.ewm({ span: 20, adjust: false }).mean();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  s.ewm({ alpha: 0.3, adjust: false }).mean();
  s.ewm({ alpha: 0.3, adjust: true }).mean();
  s.ewm({ span: 20, adjust: false }).mean();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "ewm_adjust",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_ewm_adjust.py
"""
Benchmark: EWM with adjust=False — IIR-based exponential weighted mean vs default adjust=True on 100k Series.
Outputs JSON: {"function": "ewm_adjust", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 20

data = np.sin(np.arange(SIZE) * 0.01) * 100
s = pd.Series(data)

for _ in range(WARMUP):
    s.ewm(alpha=0.3, adjust=False).mean()
    s.ewm(alpha=0.3, adjust=True).mean()
    s.ewm(span=20, adjust=False).mean()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.ewm(alpha=0.3, adjust=False).mean()
    s.ewm(alpha=0.3, adjust=True).mean()
    s.ewm(span=20, adjust=False).mean()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "ewm_adjust",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))
